### PR TITLE
Return r, s, and v without leading 0's

### PR DESCRIPTION
**Pull request by carver**
These values are [quantities, not data](https://github.com/ethereum/wiki/wiki/JSON-RPC#hex-value-encoding). Other nodes return r, s, and v as hex-encoded quantites.

See #1170

### CODE REVIEW

**Comment by frozeman** [comment]
I would actually like to get a third opinion on that. As bytes are technically invalid if they are uneven.
@chriseth, @axic any opinion?

File: packages/web3-eth-accounts/src/index.js [hunk 1]
```diff
@@ -42,14 +42,14 @@ var isNot = function(value) {
 };
 
 var trimLeadingZero = function (hex) {
-    while (hex && hex.startsWith('0x00')) {
-        hex = '0x' + hex.slice(4);
+    while (hex && hex.startsWith('0x0')) {
```
Comment (frozeman) [inline packages/web3-eth-accounts/src/index.js]: The problem with this is what if the rs, value is `0x00001234` Then a even number is made uneven

File: packages/web3-eth-accounts/src/index.js [hunk 2]
```diff
@@ -165,17 +165,17 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
 
         var rawTx = RLP.decode(rlpEncoded).slice(0, 6).concat(Account.decodeSignature(signature));
 
-        rawTx[7] = trimLeadingZero(rawTx[7]);
-        rawTx[8] = trimLeadingZero(rawTx[8]);
+        rawTx[7] = makeEven(trimLeadingZero(rawTx[7]));
+        rawTx[8] = makeEven(trimLeadingZero(rawTx[8]));
```
Comment (frozeman) [inline packages/web3-eth-accounts/src/index.js]: And are you sure passing a modified value into `RLP.decode` is not going to break anything?

**Comment by frozeman** [comment]
The tests pass here.
But this would actually be an arguement for padding the r,s values at all times with leading `0`s

**Comment by frozeman** [comment]
Additionally allowing the odd numbers again will make this problem reappear: https://github.com/ethereum/web3.js/pull/1258

Though its very important that signatures are correctly validated!

File: packages/web3-eth-accounts/src/index.js (hunk 2 above)
Comment (carver) [reply packages/web3-eth-accounts/src/index.js]: Do you mean `RLP.encode`?

It looks like it dynamically calculates the length, so I don't see any problems: https://github.com/MaiaVictor/eth-lib/blob/master/lib/rlp.js#L27

In fact, I believe leaving out 0 bytes at the beginning of an int in the RLP-encoded data is a more standard approach. (not sure if it's *required* though)

**Comment by carver** [comment]
> But this would actually be an argument for padding the r,s values at all times with leading 0s

> Additionally allowing the odd numbers again will make this problem reappear: #1258

Let's separate the two places where `r` and `s` values are used:

1. Just before rlp encoding: https://github.com/ethereum/web3.js/pull/1288/files/ac28015f159b8bece403bcac356008cc19cb037c#diff-7bb2a20126193b9ecfe4723f83429c49R168
2. When returning from `signTransaction`: https://github.com/ethereum/web3.js/pull/1288/files/ac28015f159b8bece403bcac356008cc19cb037c#diff-7bb2a20126193b9ecfe4723f83429c49R177

This PR intends to not change the functionality of # 1 - the change there was only a refactor. That means that the #1258 problem should not reappear, and I agree that `r` and `s` values should be provided with an even number of hex characters. (because that's what this rlp implementation expects)

This PR only changes the functionality of # 2, trimming all leading 0s. This brings it in line with what the nodes themselves return when returning (non-rlp-encoded) values of `r` and `s`.

**Comment by frozeman** [comment]
Ill merge this.
@malissa and @jellegerbrandy can you test the 1.0 branch when its merged?

**Comment by frozeman** [comment]
@carver this also should fix #1289, right?

**Comment by carver** [comment]
@frozeman No, some more work would have to be done, maybe in web3 or maybe in eth-lib. #1290 is the test that should pass.

Some options I see, highest preference first:

- web3 could zero-pad r and s to 32 bytes before concatenating and handing off the signature to eth-lib
- eth-lib could add a new recovery option that accepts r, s, and v as individual arguments (splitting them is the only thing it has trouble with IIRC)
- (more of a strawman than a good idea:) eth-lib could detect short r and s, maybe by doing some modified RLP decoding

**Comment by carver** [comment]
On further investigation, it looks like eth-lib itself is responsible for encoding the signature. So the best solution is to fix eth-lib's `Account.encodeSignature()`

**Comment by carver** [comment]
Submitted a PR at https://github.com/MaiaVictor/eth-lib/pull/4

That repo has 3 open pull requests that go back to October, and has never merged or closed an incoming pull request. Maybe eth-lib needs to be forked into the ethereum org to be maintained?

**Comment by frozeman** [comment]
I will talk to @MaiaVictor, but we could also leftpad this before sending it to eth-lib

**Comment by frozeman** [comment]
I merged eth-lib and the tests pass. Ill create a new version soon

**Pull request merged**